In [1]:
import torch
import torch.nn.functional as F
import torch.optim as optim
from torchvision import transforms
from session_6 import Model1, train, create_mnist_data_loaders


In [2]:
_ = torch.manual_seed(1)

In [3]:
model = Model1()
model.summarize()

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 28, 28]              80
              ReLU-2            [-1, 8, 28, 28]               0
       BatchNorm2d-3            [-1, 8, 28, 28]              16
            Conv2d-4            [-1, 8, 28, 28]             584
              ReLU-5            [-1, 8, 28, 28]               0
       BatchNorm2d-6            [-1, 8, 28, 28]              16
         MaxPool2d-7            [-1, 8, 14, 14]               0
           Dropout-8            [-1, 8, 14, 14]               0
            Conv2d-9           [-1, 12, 14, 14]             876
             ReLU-10           [-1, 12, 14, 14]               0
      BatchNorm2d-11           [-1, 12, 14, 14]              24
           Conv2d-12           [-1, 12, 14, 14]           1,308
             ReLU-13           [-1, 12, 14, 14]               0
      BatchNorm2d-14           [-1, 12,

In [4]:
epochs = 15
batch_size=128
loss_fn = F.nll_loss
optimizer = optim.Adam(model.parameters(), lr=0.01)
data_path = "../data"
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.1307,), (0.3081,))
])

In [5]:
train_loader, test_loader = create_mnist_data_loaders(batch_size, data_path, train_transform=transform, test_transform=transform)

In [6]:
train(model, epochs, train_loader, test_loader, loss_fn, optimizer)

epoch=01 loss=0.3856 batch_id=0468 accuracy=82.89%: 100%|██████████| 469/469 [00:05<00:00, 79.85it/s]


Test set: Average loss: 0.0823, Accuracy: 9773/10000 (97.73%)



epoch=02 loss=0.2334 batch_id=0468 accuracy=86.42%: 100%|██████████| 469/469 [00:04<00:00, 94.50it/s] 


Test set: Average loss: 0.0481, Accuracy: 9858/10000 (98.58%)



epoch=03 loss=0.2942 batch_id=0468 accuracy=86.71%: 100%|██████████| 469/469 [00:04<00:00, 94.24it/s]


Test set: Average loss: 0.0771, Accuracy: 9764/10000 (97.64%)



epoch=04 loss=0.3171 batch_id=0468 accuracy=87.06%: 100%|██████████| 469/469 [00:04<00:00, 96.43it/s] 


Test set: Average loss: 0.0352, Accuracy: 9883/10000 (98.83%)



epoch=05 loss=0.3700 batch_id=0468 accuracy=86.86%: 100%|██████████| 469/469 [00:04<00:00, 94.69it/s]


Test set: Average loss: 0.0338, Accuracy: 9890/10000 (98.90%)



epoch=06 loss=0.2911 batch_id=0468 accuracy=87.37%: 100%|██████████| 469/469 [00:05<00:00, 93.10it/s]


Test set: Average loss: 0.0241, Accuracy: 9924/10000 (99.24%)



epoch=07 loss=0.3265 batch_id=0468 accuracy=87.39%: 100%|██████████| 469/469 [00:04<00:00, 95.16it/s] 


Test set: Average loss: 0.0256, Accuracy: 9919/10000 (99.19%)



epoch=08 loss=0.3459 batch_id=0468 accuracy=87.41%: 100%|██████████| 469/469 [00:04<00:00, 97.07it/s] 


Test set: Average loss: 0.0247, Accuracy: 9920/10000 (99.20%)



epoch=09 loss=0.2328 batch_id=0468 accuracy=87.54%: 100%|██████████| 469/469 [00:04<00:00, 94.66it/s]


Test set: Average loss: 0.0308, Accuracy: 9895/10000 (98.95%)



epoch=10 loss=0.3972 batch_id=0468 accuracy=87.66%: 100%|██████████| 469/469 [00:04<00:00, 95.62it/s]


Test set: Average loss: 0.0281, Accuracy: 9905/10000 (99.05%)



epoch=11 loss=0.2075 batch_id=0468 accuracy=87.80%: 100%|██████████| 469/469 [00:05<00:00, 93.18it/s]


Test set: Average loss: 0.0271, Accuracy: 9918/10000 (99.18%)



epoch=12 loss=0.2350 batch_id=0468 accuracy=87.76%: 100%|██████████| 469/469 [00:05<00:00, 93.70it/s] 


Test set: Average loss: 0.0269, Accuracy: 9916/10000 (99.16%)



epoch=13 loss=0.3078 batch_id=0468 accuracy=87.55%: 100%|██████████| 469/469 [00:04<00:00, 94.76it/s] 


Test set: Average loss: 0.0254, Accuracy: 9921/10000 (99.21%)



epoch=14 loss=0.3115 batch_id=0468 accuracy=87.88%: 100%|██████████| 469/469 [00:05<00:00, 92.61it/s]


Test set: Average loss: 0.0229, Accuracy: 9922/10000 (99.22%)



epoch=15 loss=0.2397 batch_id=0468 accuracy=87.67%: 100%|██████████| 469/469 [00:04<00:00, 98.15it/s] 


Test set: Average loss: 0.0237, Accuracy: 9920/10000 (99.20%)



**Target**

Target was to build the skeleton, add all proper layers, and get a model below 8k parameters

**Result**

The skeleton was taken from session 5 assignment and modified a bit. Instead of a direct jump from 8 channels to 16 channels between block 1 and 2, it is first increased to 12 channels and then 16 (in block 3).

Parameter Count: 6174

train accuracy = 87.67%

test accuracy = 99.20%

**Analysis**

Good structure, parameter count is below target. But model is underfitting. Need to reduce regularization.
Both test and train accuracies are not stable.